tiledb/sm/filesystem/vfs.h

```diff
@@ -348,6 +356,16 @@ class VFS {
    */
   Status ls(const URI& parent, std::vector<URI>* uris) const;
 
+  /**
+   * Recursively lists objects and object information that start with `prefix`.
+   *
+   * @param prefix The parent path to list sub-paths.
+   * @param uris The URIs that are contained in the parent.
+   * @param max_paths The maximum number of paths to be retrieved.
+   */
+  void ls_recursive(
```

Comment:
Should we have a `optional<std::vector<URI>>` return instead of pass the vector as a pointer?

Reply:
Yep that makes sense. These will throw on failure though, unless I'm mistaken we'd never return the nullopt so I updated to return `vector<URI>`.